RDRP-1172_construction

Ryan2Y79 opened the pull request:
## Pull Request submission

Construction hotfix. The addition of this line in a previous PR was totally redundant, and did not have the intended effect as to whether NI construction was run or not. Furthermore, the addition of this line is causing the pipeline to die during construction. 

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #


#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [x] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [x] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [x] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [x] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [x] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)


#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

Ryan2Y79 committed:
Update pipeline.py

adriano-lopresti reviewed (commented):
i just had a quick look at it, but my questions are:
1) it's just one line you've added?
2) what you added with that line, is adding one argument when the function run_constructions is called.. and i notice that the function is already defined with that argument (but it's an optional value, the default being False)..
3) i see that, in the config, the value of that parameter is False, which is the same as the default parameter
 
so the question is: what is the context of this? it seems that effectively, with this change nothing as changed.
Except that it gives the possibility of switching that parameter from False to True. Have you tested what happen if this is switched to True? I checked that the pipeline still runs through, but i haven't verified the impact of doing so.
 
But i still find the logic of it all a bit puzzling
(but what i find puzzling was actually there before your PR...your PR gives the possibility of "switching off" this chunk of code.. so maybe one should jsut remove this line 56..)
What i find puzzling is: to me is seems that the variable run_construction defined in line line 56 of contruction_main is redundant 
You pass a "is_northern_ireland" argument(which is read from the run_ni_contruction config parameter).
Then if this true, you define a variable run_construction being equal to the same config parameter (why?) which doesn't seem to be subsequenrly used.

Ryan2Y79 commented:
The change I prepose here is reversing a line I added in this PR (that you approved and merged) https://github.com/ONSdigital/research-and-development/pull/409/files. 

The overall context is that Northern Ireland and GB construction should happen independently. Norther Ireland construction happens `pipeline.py` lines 139-153, and GB construction happens in `pipeline.py` lines 155-168. However BOTH construction executions use the same construction function `run_construction`. 

The reason for the `is_northern_ireland` parameter and its default value to set to `False` is that when GB construction is run, we do NOT want NI construction to happen. And when GB constructions happens, we do NOT want NI construction to happen. Both GB and NI construction should happen happen independently. My change in PR `https://github.com/ONSdigital/research-and-development/pull/409/files` violates this idea by having Northern Ireland construction happen at the same time as GB if config["global"]["run_ni_construction"] is set to `True`

You are totally correct in that with the config["global"]["run_ni_construction"] set to False, my change here makes no difference. But when config["global"]["run_ni_construction"] is set to True, the pipeline will die as we have introduced NI construction during GB construction. 

In short, I want to reverse a change that I made previously to restore the pipeline to how it once was.

adriano-lopresti reviewed (approved):
I approve
